### PR TITLE
Add Python 3.10 testing.

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

Currently, pillow (https://github.com/python-pillow/Pillow/issues/5569) and kiwisolver (https://github.com/nucleic/kiwi/pull/103) do not have wheels, which slows down the build a bit, but it should pass anyway.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [n/a] New features are documented, with examples if plot related.
- [n/a] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [n/a] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [n/a] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by QuLogic** [comment]
Oh, looks like we also need tornado, pandas, pikepdf, pandocfilters, and pyrsistent wheels, but pikepdf is the one tripping us up due to its C libraries.

**Comment by QuLogic** [comment]
I opened pikepdf/pikepdf#245, but it might be a while since Pillow doesn't have wheels either.

**Comment by QuLogic** [comment]
So the failures are now:

- [ ] Qt5 with PySide2: TypeError: 'PySide2.QtCore.Qt.ToolBarArea' object cannot be interpreted as an integer
- [ ] `test_fork`, `test_savefig_backend`, `test_load_from_url`, `test_use_url`: ImportWarning: DynamicImporter.find_spec() not found; falling back to find_module()` -> https://gitlab.gnome.org/GNOME/pygobject/-/issues/473
- [ ] `test_tinypages`: ImportError: cannot import name 'Union' from 'types' -> https://github.com/sphinx-doc/sphinx/issues/9562

**Review by jkseppan — APPROVED** [review]
Is it possible to make the 3.10 test optional, so it doesn't cause a red "Some checks were not successful" message on Github?

**Comment by dopplershift** [comment]
GitHub actions doesn't have a notion of "allowed failures"--every PR workflow shows up with it's status here. The options are:
* Manually override and mark as passing (seems silly)
* Make the Python 3.10/pre-release build something that runs periodically on a cron (e.g. nightly). For this one it's possible to have the workflow open an issue if it fails.

**Comment by tacaswell** [comment]
We could forgive the ImportWarning's in the configuration (like we are doing with deprecation warnings from fontTools right now (and fontTools fixed that upstream yesterday))?

I'll look at the pyside2 issue and the sphinx issue as already fixed on their default branch.

**Comment by QuLogic** [comment]
> We could forgive the ImportWarning's in the configuration (like we are doing with deprecation warnings from fontTools right now (and fontTools fixed that upstream yesterday))?

Oh yea, we could do that.

> I'll look at the pyside2 issue and the sphinx issue as already fixed on their default branch.

I have a work around for PySide2 mostly complete.

**Comment by tacaswell** [comment]
> I have a work around for PySide2 mostly complete.

Great, I was just about to start on this and now will not :)

File: lib/matplotlib/backends/backend_qt.py [hunk 1]
```diff
@@ -626,9 +626,9 @@ class NavigationToolbar2QT(NavigationToolbar2, QtWidgets.QToolBar):
     def __init__(self, canvas, parent, coordinates=True):
         """coordinates: should we show the coordinates on the right?"""
         QtWidgets.QToolBar.__init__(self, parent)
-        self.setAllowedAreas(
-            _enum("QtCore.Qt.ToolBarArea").TopToolBarArea
-            | _enum("QtCore.Qt.ToolBarArea").TopToolBarArea)
+        self.setAllowedAreas(QtCore.Qt.ToolBarAreas(
+            int(_enum("QtCore.Qt.ToolBarArea").TopToolBarArea) |
+            int(_enum("QtCore.Qt.ToolBarArea").BottomToolBarArea)))
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_qt.py:631]: Note the previous typo in the refactor of Qt5/Qt6 code together.

File: lib/matplotlib/backends/backend_qt.py [hunk 2]
```diff
@@ -883,13 +883,13 @@ class ToolbarQt(ToolContainerBase, QtWidgets.QToolBar):
     def __init__(self, toolmanager, parent):
         ToolContainerBase.__init__(self, toolmanager)
         QtWidgets.QToolBar.__init__(self, parent)
-        self.setAllowedAreas(
-            _enum("QtCore.Qt.ToolBarArea").TopToolBarArea
-            | _enum("QtCore.Qt.ToolBarArea").TopToolBarArea)
+        self.setAllowedAreas(QtCore.Qt.ToolBarAreas(
+            int(_enum("QtCore.Qt.ToolBarArea").TopToolBarArea) |
+            int(_enum("QtCore.Qt.ToolBarArea").BottomToolBarArea)))
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_qt.py:888]: Ditto.

**Review by QuLogic — COMMENTED** [review]
The change to enums is enough to work around PySide2 in tests for me, but if I actually run an example, it crashes in a NumPy copy during a paint operation. It doesn't happen in TkAgg or PyQt5, so I'm not sure what I can do about it.

**Review by tacaswell — APPROVED** [review]

**Pull request merged**